src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Unix.cs

```diff
@@ -61,5 +61,98 @@ internal static unsafe void GetAnsiStringBytes(ReadOnlySpan<char> chars, Span<by
             int actualByteLength = Encoding.UTF8.GetBytes(chars, bytes);
             bytes[actualByteLength] = 0;
         }
+
+        public static IntPtr AllocHGlobal(IntPtr cb)
+        {
+            nuint cbNative = (nuint)(nint)cb;
+
+            // Avoid undefined malloc behavior by always allocating at least one byte
+            IntPtr pNewMem = Interop.Sys.MemAlloc((cbNative != 0) ? cbNative : 1);
+            if (pNewMem == IntPtr.Zero)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            return pNewMem;
+        }
+
+        public static void FreeHGlobal(IntPtr hglobal)
+        {
+            if (hglobal != IntPtr.Zero)
+            {
+                Interop.Sys.MemFree(hglobal);
+            }
+        }
+
+        public static IntPtr ReAllocHGlobal(IntPtr pv, IntPtr cb)
+        {
+            nuint cbNative = (nuint)(nint)cb;
+
+            // Match Windows behavior by always allocating at least one byte and only when the original pointer is non-zero
+            IntPtr pNewMem = (pv != IntPtr.Zero) ? Interop.Sys.MemReAlloc(pv, (cbNative != 0) ? cbNative : 1) : IntPtr.Zero;
+            if (pNewMem == IntPtr.Zero)
+            {
+                throw new OutOfMemoryException();
+            }
+            return pNewMem;
+        }
+
+        public static IntPtr AllocCoTaskMem(int cb) => AllocHGlobal((nint)(uint)cb);
+
+        public static void FreeCoTaskMem(IntPtr ptr) => FreeHGlobal(ptr);
+
+        public static IntPtr ReAllocCoTaskMem(IntPtr pv, int cb)
+        {
+            nuint cbNative = (nuint)(uint)cb;
+
+            if (cbNative == 0)
+            {
+                if (pv != IntPtr.Zero)
+                {
+                    Interop.Sys.MemFree(pv);
+                    return IntPtr.Zero;
+                }
+                // Avoid undefined realloc behavior by always allocating at least one byte
+                cbNative = 1;
+            }
+
+            IntPtr pNewMem = Interop.Sys.MemReAlloc(pv, cbNative);
+            if (pNewMem == IntPtr.Zero)
+            {
+                throw new OutOfMemoryException();
+            }
+            return pNewMem;
+        }
+
+        internal static unsafe IntPtr AllocBSTR(int length)
+        {
+            // SysAllocString on Windows aligns the memory block size up
+            const nuint WIN32_ALLOC_ALIGN = 15;
+
+            ulong cbNative = 2 * (ulong)(uint)length + (uint)sizeof(IntPtr) + (uint)sizeof(char) + WIN32_ALLOC_ALIGN;
+            if (cbNative > uint.MaxValue)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            IntPtr p = Interop.Sys.MemAlloc((nuint)cbNative & ~WIN32_ALLOC_ALIGN);
+            if (p == IntPtr.Zero)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            IntPtr s = p + sizeof(IntPtr);
+            *(((uint*)s) - 1) = (uint)(length * sizeof(char));
+
+            return s;
```

Comment:
(not sure if it is worth it but) to keep the 32-bit support as before, could it be:

```suggestion
#if TARGET_64BIT
            p += sizeof(IntPtr);
#endif
            *(((uint*)p) - 1) = (uint)(length * sizeof(char));

            return p;
```

Reply:
This is not what the BSTR allocation did before this change. In fact, this edit would introduce buffer overrun on 32-bit platforms.

Reply:
I was interpreting this block: https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/coreclr/src/palrt/bstr.cpp#L86-L94

Reply:
It looks correct to me. For reference, the runtime code in Mono (which I fixed up recently): https://github.com/dotnet/runtime/blob/bac8819d96cd40cb45d0bc6c0c9f33489c64beec/src/mono/mono/metadata/cominterop.c#L3005-L3009